### PR TITLE
AP_Motors: refactor motor matrix to cleanup and make symbols

### PR DESCRIPTION
**Pull request by Hwurzburg**
Allows searching for frame type build options in the ELF file

### CODE REVIEW

File: libraries/AP_Motors/AP_MotorsMatrix.cpp [hunk 1]
```diff
@@ -63,26 +63,26 @@ bool AP_MotorsMatrix::init(uint8_t expected_num_motors)
...
+        _mav_type = MAV_TYPE_OCTOROTOR;
+        break;
+    case 10:
+        _mav_type = MAV_TYPE_DECAROTOR;
+        break;
+    case 12:
+        _mav_type = MAV_TYPE_DODECAROTOR;
+        break;
+    default:
+        _mav_type = MAV_TYPE_GENERIC;
```
Comment (IamPete1) [inline libraries/AP_Motors/AP_MotorsMatrix.cpp:85]: I'm not a fan of this change.

File: libraries/AP_Motors/AP_MotorsMatrix.cpp [hunk 2]
```diff
@@ -145,33 +145,33 @@ void AP_MotorsMatrix::output_to_motors()
...
+    case SpoolState::THROTTLE_UNLIMITED:
+    case SpoolState::SPOOLING_DOWN:
+        // set motor output based on thrust requests
+        for (i = 0; i < AP_MOTORS_MAX_NUM_MOTORS; i++) {
+            if (motor_enabled[i]) {
+                set_actuator_with_slew(_actuator[i], thrust_to_actuator(_thrust_rpyt_out[i]));
             }
-            break;
+        }
+        break;
```
Comment (IamPete1) [inline libraries/AP_Motors/AP_MotorsMatrix.cpp:174]: again not a fan of the above white space changes.

File: libraries/AP_Motors/AP_MotorsMatrix.cpp [hunk 3]
```diff
@@ -569,6 +569,645 @@ void AP_MotorsMatrix::add_motors_raw(const struct MotorDefRaw *motors, uint8_t n
...
+        success = false;
+        break;
+    } //hexa
+    return success;
+}
+#endif ////AP_MOTORS_FRAME_HEXA_ENABLED
+#if AP_MOTORS_FRAME_OCTA_ENABLED
+bool AP_MotorsMatrix::setup_octa_matrix(motor_frame_type frame_type)
+{
+    bool success = true;
```
Comment (IamPete1) [inline libraries/AP_Motors/AP_MotorsMatrix.cpp:855]: don't need this bool, the same for all of them.

File: libraries/AP_Motors/AP_MotorsMatrix.cpp [hunk 4]
```diff
@@ -569,6 +569,645 @@ void AP_MotorsMatrix::add_motors_raw(const struct MotorDefRaw *motors, uint8_t n
...
+            {  -22.5f,  AP_MOTORS_MATRIX_YAW_FACTOR_CW,   8 },
+        };
+        add_motors(motors, ARRAY_SIZE(motors));
+        break;
+    }
+    default:
+        // octa frame class does not support this frame type
+        _frame_type_string = "UNSUPPORTED";
+        success = false;
+        break;
```
Comment (IamPete1) [inline libraries/AP_Motors/AP_MotorsMatrix.cpp:967]:
```suggestion
        return false;
```

File: libraries/AP_Motors/AP_MotorsMatrix.cpp [hunk 5]
```diff
@@ -569,6 +569,645 @@ void AP_MotorsMatrix::add_motors_raw(const struct MotorDefRaw *motors, uint8_t n
...
+    } // octa frame type
+    return success;
```
Comment (IamPete1) [inline libraries/AP_Motors/AP_MotorsMatrix.cpp:969]:
```suggestion
    return true;
```

File: libraries/AP_Motors/AP_MotorsMatrix.cpp [hunk 6]
```diff
@@ -581,645 +1220,60 @@ void AP_MotorsMatrix::setup_motors(motor_frame_class frame_class, motor_frame_ty
...
-                default:
-                    // quad frame class does not support this frame type
-                    _frame_type_string = "UNSUPPORTED";
-                    success = false;
-                    break;
-            }
-            break;  // quad
+    case MOTOR_FRAME_QUAD:
+        _frame_class_string = "QUAD";
+        _mav_type = MAV_TYPE_QUADROTOR;
```
Comment (IamPete1) [inline libraries/AP_Motors/AP_MotorsMatrix.cpp:1225]: why not move these two down into the new method? (same for all)

File: libraries/AP_Motors/AP_MotorsMatrix.h [hunk 7]
```diff
@@ -151,5 +161,7 @@ class AP_MotorsMatrix : public AP_MotorsMulticopter {
     const char*         _frame_class_string = ""; // string representation of frame class
     const char*         _frame_type_string = "";  //  string representation of frame type
 private:
+    
+
```
Comment (IamPete1) [inline libraries/AP_Motors/AP_MotorsMatrix.h:165]: unneeded white space change.

**Review by IamPete1 — CHANGES_REQUESTED** [review]
Costing more flash than I would have guessed, I suppose the whole point was to add symbols to the elf....

File: libraries/AP_Motors/AP_MotorsMatrix.h (hunk 7 above)
Comment (Hwurzburg) [reply libraries/AP_Motors/AP_MotorsMatrix.h:165]: removed

**Comment by Hwurzburg** [comment]
rebased after @andyp1per recent changes

**Review by IamPete1 — APPROVED** [review]
LGTM, thanks

**Review by rmackay9 — APPROVED** [review]

**Pull request merged**